### PR TITLE
Refactor parsing of Dockerfile into separate functions

### PR DESCRIPTION
**Pull request by ashb**
Rather than a single function that parses both the preamble and the main
body this splits the parsing into two distinct phases.

And since Go has no native "iterator" type that we can pass between functions
we use a channel that yields the tokens

### CODE REVIEW

File: buildkit/internal/transform/transforms.go [hunk 1]
```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
...
-
-				// Try to use the astro-runtime base (non-onbuild) image. If we have any error, "fail safe"
-				// and leave it unmodified.
-				_ = ensureValidBaseImage(node, buildArgs)
-				adjustedLine = -node.EndLine
-				inPreamble = false
-			case pyenvCommand:
-				return nil, nil, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
+
+tokens:
```
Comment (ianbuss) [inline buildkit/internal/transform/transforms.go:82]: nit: can we call the label something else? Having it the same name as an arg is a bit confusing.

File: buildkit/internal/transform/transforms.go [hunk 2]
```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
...
+				break tokens
+			case tokenStream <- node:
+				continue
+			}
+		}
+		close(tokenStream)
+	}(ctx)
+	defer cancel()
+
+	premable, adjustedLine, err := transformer.parsePreamble(tokenStream)
```
Comment (ianbuss) [inline buildkit/internal/transform/transforms.go:153]:
```suggestion
	preamble, adjustedLine, err := transformer.parsePreamble(tokenStream)
```

File: buildkit/internal/transform/transforms.go [hunk 3]
```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
...
+	if err != nil {
+		return nil, nil, err
+	}
+	transformedAst, err := transformer.parseMain(tokenStream, adjustedLine)
+
+	return premable, transformedAst, err
```
Comment (ianbuss) [inline buildkit/internal/transform/transforms.go:159]:
```suggestion
	return preamble, transformedAst, err
```

File: buildkit/internal/transform/transforms.go [hunk 4]
```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
...
+			adjustedLine = -node.EndLine
+			break tokens
+		case pyenvCommand:
+			return nil, 0, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
 		}
+	}
+	return preamble, adjustedLine, nil
+}
+
+func (transformer *Transformer) parseMain(tokens chan *parser.Node, lineNoAdjustment int) (*parser.Node, error) {
```
Comment (ianbuss) [inline buildkit/internal/transform/transforms.go:105]:
```suggestion
func (transformer *Transformer) parseMain(tokens chan *parser.Node, lineNumAdjustment int) (*parser.Node, error) {
```
Just to be clear it doesn't mean no adjustment

File: buildkit/internal/transform/transforms.go [hunk 5]
```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
...
+	if err != nil {
+		return nil, nil, err
+	}
+	transformedAst, err := transformer.parseMain(tokenStream, adjustedLine)
+
+	return premable, transformedAst, err
 }
 
-func processArg(node *parser.Node, buildArgs map[string]string) error {
+func (r *Transformer) processArg(node *parser.Node) error {
```
Comment (ianbuss) [inline buildkit/internal/transform/transforms.go:162]: Go style guidelines say to use the same receiver var name in all functions. I think rename all to `t`?

**Review by ianbuss — COMMENTED** [review]
Couple of comments

File: buildkit/internal/transform/transforms.go (hunk 5 above)
Comment (ashb) [reply buildkit/internal/transform/transforms.go:162]: Yes to `t` over `r` -- I wasn't sure if everything should be on the Transformer or if some of the functions could/should stay at top level

Comment (ianbuss) [reply buildkit/internal/transform/transforms.go:162]: I would say use the same receiver everywhere and whether the function is exported implies it is top level? I may not have grasped what you mean there though.

Comment (ashb) [reply buildkit/internal/transform/transforms.go:162]: Yeah, I guess in python this one would be a "static" function -- and go doesn't have/need such a concept, so this is currently just a "helper function".

Can make it a struct method for consistency though

Comment (ashb) [reply buildkit/internal/transform/transforms.go:162]: Thinking a little bit more about it -- it seems silly to make `validateVirtualEnv` and `extractPythonMajorMinor` have the receiver too

Comment (ashb) [reply buildkit/internal/transform/transforms.go:162]: Oh. Yeah I see your point. I missed changing one, I was thinking this diff comment was a suggestion!

**Review by ianbuss — APPROVED** [review]
Just last nits

**Pull request merged**